sys/log/full/src/log_fcb.c

```diff
@@ -97,46 +100,185 @@ log_fcb_append(struct log *log, void *buf, int len)
+     * many bytes are required to increase the chunk size up to a multiple of
+     * the flash alignment).
+     */
+    hdr_alignment = log_fcb_first_chunk_sz(fcb->f_align);
+    if (hdr_alignment > body_len) {
+        chunk_sz = sizeof *hdr + body_len;
+    } else {
+        chunk_sz = sizeof *hdr + hdr_alignment;
     }
 
-    rc = log_fcb_start_append(log, len, &loc);
-    if (rc) {
-        goto err;
+    u8p = body;
+
+    memcpy(buf, hdr, sizeof *hdr);
+    memcpy(buf + sizeof *hdr, u8p, hdr_alignment);
+
+    rc = flash_area_write(loc.fe_area, loc.fe_data_off, buf, chunk_sz);
+    if (rc != 0) {
+        return rc;
     }
 
+    /* Append the remainder of the message body. */
+
+    u8p += hdr_alignment;
+    body_len -= hdr_alignment;
+
+    if (body_len > 0) {
+        rc = flash_area_write(loc.fe_area, loc.fe_data_off + chunk_sz, u8p,
+                              body_len);
+        if (rc != 0) {
+            return rc;
+        }
+    }
+
+    rc = fcb_append_finish(fcb, &loc);
+    if (rc != 0) {
+        return rc;
+    }
+
+    return 0;
+}
+
+static int
+log_fcb_write_mbuf(struct fcb_entry *loc, const struct os_mbuf *om)
+{
+    int rc;
+
     while (om) {
-        rc = flash_area_write(loc.fe_area, loc.fe_data_off, om->om_data,
-                              om->om_len);
-        if (rc) {
-            goto err;
+        rc = flash_area_write(loc->fe_area, loc->fe_data_off,
+                              om->om_data, om->om_len);
+        if (rc != 0) {
+            return SYS_EIO;
         }
 
-        loc.fe_data_off += om->om_len;
+        loc->fe_data_off += om->om_len;
         om = SLIST_NEXT(om, om_next);
     }
 
+    return 0;
+}
+
+static int
+log_fcb_append_mbuf(struct log *log, const struct os_mbuf *om)
+{
+    struct fcb *fcb;
+    struct fcb_entry loc;
+    struct fcb_log *fcb_log;
+    int len;
+    int rc;
+
+    fcb_log = (struct fcb_log *)log->l_arg;
+    fcb = &fcb_log->fl_fcb;
+
+    /* This function expects to be able to write each mbuf without any
+     * buffering.
+     */
+    if (fcb->f_align != 1) {
+        return SYS_ENOTSUP;
+    }
+
+    len = os_mbuf_len(om);
+    rc = log_fcb_start_append(log, len, &loc);
+    if (rc != 0) {
+        return rc;
+    }
+
+    rc = log_fcb_write_mbuf(&loc, om);
+    if (rc != 0) {
+        return rc;
+    }
+
     rc = fcb_append_finish(fcb, &loc);
+    if (rc != 0) {
+        return rc;
+    }
 
-err:
-    return (rc);
+    return 0;
+}
+
+static int
+log_fcb_append_mbuf_body(struct log *log, const struct log_entry_hdr *hdr,
+                         const struct os_mbuf *om)
+{
+    struct fcb *fcb;
+    struct fcb_entry loc;
+    struct fcb_log *fcb_log;
+    int len;
+    int rc;
+
+    fcb_log = (struct fcb_log *)log->l_arg;
+    fcb = &fcb_log->fl_fcb;
+
+    /* This function expects to be able to write each mbuf without any
+     * buffering.
+     */
+    if (fcb->f_align != 1) {
```

Comment:
Maybe a /* XXXX */ comment here to call attention in case someone ends up missing this, and wondering where to start add support for it :)